### ISSUE TITLE
Changing pulumi name of aws.s3.Bucket with same bucket name fails because names must be unique and it tries create-before-delete

### ISSUE THREAD

**Issue by gobengo** [pulumi/pulumi-aws#511]
I have only been using Pulumi for a few days, but so far so good.

I am setting up a fresh set of AWS resources with Route53, S3, and Cloudfront to set up a redirect from 'conserve.org' to 'www.conserve.org'.

As I iterate, these resources often get recreated. But some AWS Resources have uniqueness constraints. For example, s3 bucket names have to be unique, and what I'm doing requires that the s3 bucket name be the same as the apex domain (i.e. 'conserve.org').

At first I had some janky *pulumi component* names that led to collisions, but now I've ironed that out. But when I do `pulumi up`, pulumi tries to
1. Create a new S3 Bucket with the same S3 bucket name (and a different pulumi state name)
2. Then delete the old s3 bucket with that name and the now-not-required pulumi name

But it fails at step 1 because AWS rejects the request to create a new S3 Bucket because the name is already taken.

I can see that lower level work on [Delete Before Create](https://github.com/pulumi/pulumi/pull/1365) is relatively recent. Has it just not been used by the pulumi-aws library? Do I need to do something special to avoid what's happening?

Thanks!

**Comment by lukehoban** [comment]
If this was truly a replacement, you should be able to pass `deleteBeforeReplace: true` to allow this replacement to take affect. Note that this will result in a downtime where no bucket exists - but I assume that is okay in your case. 

However, if I understand your case correctly, you have logically two distinct Pulumi resources, which are both tied to the same bucket name. The one you previously created, and the one you are trying to create. Pulumi doesn’t see this as a replacement, it sees it as two different resources, one of which is being deleted and the other created. Pulumi also always tries to process creates before unrelated deletes. 

There’s a few ways you could handle this currently:
1. Do this as two separate updates - first one that removes it (without the code present to create the new one) and then one that creates it. 
2. `pulumi stack export` and then modify the state file to give the resource the new Pulumi URN, then import back in. Then it will be seen as a noop.

We’re working on some tooling to make renaming a resource as in (2) easier, as well as tooling to “adopt” existing resources which is also related to (2).

**Comment by gobengo** [comment]
Thanks for the quick answer

> If this was truly a replacement, you should be able to pass deleteBeforeReplace: true to allow this replacement to take affect.

Am I understanding this right to imply that if I ever want the engine to do delete-before-replace, I need to opt-in by passing deleteBeforeReplace?

Will there ever be a world where the pulumi client 'knows' about things like S3's name uniqueness constraint and automatically does deleteBeforeReplace without me (or a teammate) having to know these quirks of AWS? I would understand if not (nothing else I've used ever has), but it would be *amazing* if it did.

As a small aside: I had a hard time finding documentation for this (but did search for a bit) Because all the github issues talk about 'delete-before-create', but the end-user API exposes `deleteBeforeReplace`. Thanks to your tip I have found [the docs](https://pulumi.io/reference/programming-model.html#deletebeforereplace)

> However, if I understand your case correctly, you have logically two distinct Pulumi resources, which are both tied to the same bucket name.

Either you're misunderstanding or 'logically distinct' is in the eye of the beholder. From my developer experience, I just changed a string as the first parameter of the `Bucket` constructor to better generate unique names (or even just to change the name for aesthetic reasons). It's not really a 'logical change' to me. But I think I understand more about the programming model to get what I want.

> 2. pulumi stack export and then modify the state file to give the resource the new Pulumi URN, then import back in. Then it will be seen as a noop.

Got it. That's how I can 'rename' things now. I would have been worried this might break references in the component hierarchy or something. Good to know (it sounds like) I'm safe in just changing the name in the URN to whatever I want.
